src/yieldOptions/providers/protocols/compoundv3.ts

```diff
@@ -0,0 +1,208 @@
+                    // @ts-ignore
+                    rewardToken: CompAddress[chainId],
+                    apy: Number(supplyCompRewardApr),
+                }
+            ]
+
+        };
+        return result;
+    }
+
+    getAssets(chainId: number): Promise<Address[]> {
+        if (chainId !== 1) throw new Error("Compound V2 is only available on Ethereum mainnet");
+
+        return Promise.resolve(Object.keys(assetToCToken[chainId]).map((key) => key as Address));
+    }
+
+
+    private getBaseScale(client: PublicClient, cToken: Address): Promise<bigint> {
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "baseScale",
+        });
+    }
+
+    private getUtilization(client: PublicClient, cToken: Address): Promise<bigint> {
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "getUtilization",
+        });
+    }
+    private getSupplyRate(client: PublicClient, cToken: Address, utilization: bigint): Promise<bigint> {
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "getSupplyRate",
+            args: [utilization],
+        });
+    }
+    private getTotalSupply(client: PublicClient, cToken: Address): Promise<bigint> {
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "totalSupply",
+        });
+    }
+    private getBaseTrackingSupplySpeed(client: PublicClient, cToken: Address): Promise<bigint> {
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "baseTrackingSupplySpeed",
+        });
+    }
+
+    private getBaseTokenPriceFeed(client: PublicClient, cToken: Address): Promise<Address> {
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "baseTokenPriceFeed",
+        });
+    }
+    private getPrice(client: PublicClient, cToken: Address, priceFeed: Address): Promise<bigint> {
+        // baseTokenPriceFeed for cWETH on mainnet is a constant price oracle which just returns 1e6.
+        // So we need to use the Chainlink WETH/USD oracle here
+        if (cToken === "0xA17581A9E3356d9A858b789D68B4d866e593aE94") {
+            return this.getEthPrice();
+        }
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "getPrice",
+            args: [priceFeed],
+        });
+    }
+
+    private getTrackingIndexScale(client: PublicClient, cToken: Address): Promise<bigint> {
+        return client.readContract({
+            address: cToken,
+            abi: CTOKEN_ABI,
+            functionName: "trackingIndexScale"
+        });
+    }
+
+    private async getCompPrice(): Promise<bigint> {
+        if (!this.clients[1]) {
```

Comment:
Can we let the user know somehow that they have to always add a mainnet client?

Reply:
I've moved the check to the constructor. That way they will get an error message early on.